### PR TITLE
Grafting of subgraphs

### PR DESCRIPTION
**Pull request by lutter**
This PR implements subgraph grafting as described in this [engineering plan](https://github.com/graphprotocol/rfcs/blob/master/engineering-plans/0004-subgraph-grafting.md)

### CODE REVIEW

File: core/src/subgraph/registrar.rs [hunk 1]
```diff
@@ -466,18 +466,26 @@ fn create_subgraph(
...
     ethereum_adapter: Arc<dyn EthereumAdapter>,
     logger: &Logger,
 ) -> Box<
     dyn Future<
-            Item = (Option<EthereumBlockPointer>, Option<EthereumBlockPointer>),
+            Item = (
+                Option<EthereumBlockPointer>,
+                Option<EthereumBlockPointer>,
+                Option<EthereumBlockPointer>,
+            ),
```
Comment (leoyvens) [inline core/src/subgraph/registrar.rs:482]: We should introduce a struct for this.

File: graph/src/data/subgraph/mod.rs [hunk 2]
```diff
@@ -854,6 +892,8 @@ pub struct BaseSubgraphManifest<S, D, T> {
     pub repository: Option<String>,
     pub schema: S,
     pub data_sources: Vec<D>,
+    #[serde(rename = "graft")]
+    pub base: Option<Graft>,
```
Comment (leoyvens) [inline graph/src/data/subgraph/mod.rs:896]: Why not have the field name match the serialized name?

File: core/src/subgraph/registrar.rs (hunk 1 above)
Comment (lutter) [reply core/src/subgraph/registrar.rs:482]: I thought about that, but the function is used only once, in `create_subgraph_version`, and really is just a detail of that function. We should just rewrite `resolve_subgraph_chain_blocks` as three async functions and then `await` them in `create_subgraph_version`.

File: graph/src/data/subgraph/mod.rs (hunk 2 above)
Comment (lutter) [reply graph/src/data/subgraph/mod.rs]: I am sure I had a very good reason - just changed it to `graft`

File: store/postgres/src/relational.rs [hunk 3]
```diff
@@ -185,12 +185,12 @@ impl Layout {
                 TypeDefinition(Interface(_)) => { /* we do not care about interfaces */ }
                 TypeDefinition(Enum(enum_type)) => {
                     SqlName::check_valid_identifier(&enum_type.name, "enum")?;
-                    let values: Vec<_> = enum_type
+                    let values: HashSet<_> = enum_type
```
Comment (leoyvens) [inline store/postgres/src/relational.rs:188]: It's better to use a `BTreeSet` here so we have a defined order.

**Comment by lutter** [comment]
Rebased to latest master and addressed all comments so far

**Comment by lutter** [comment]
Reworked so that each commit passes `cargo check` individually

**Comment by lutter** [comment]
I just tried this with a fairly large production subgraph (`ianlapham/uniswap-hardcoded`, `QmdmdZRG42GG4C4JWNCnf5uwGAG8Uh6Q6iAM1VDLnWKNzq`) which weighs in at around 50GB, and the deploy took about 4 hours and 45 minutes on a test database. The test database is much slower than prod, but even in prod this would be a significant delay.

The code needs to be changed to do the copying asynchronously. What should happen is that during deploy the schemas are checked for compatibility using `Layout.can_copy_from`, but that the actual copying done by `Layout.copy_from` doesn't happen until the subgraph is actually started. That will be triggered by `entities::Connection.start_subgraph` We'll (ab) use the existing column `deployment_schemas.state` to signal that: a normal subgraph has state `Ready` here, we will add a state `Initializing`. Deploying a subgraph will set the state to `Initializing`, and `start_subgraph` will transition from `Initializing` -> `Ready` by calling `Layout.copy_from` if the subgraph has a graft set.

Since copying can take a long time, and there is no good way to indicate progress, we should log something like `Starting to copy FooBar entities` and `Copied FooBar entities in 439s` on the subgraph's logger.

**Comment by lutter** [comment]
Moved data copying to the background (during `start_subgraph`) and rebased on top of latest master.

**Comment by lutter** [comment]
> I just tried this with a fairly large production subgraph (`ianlapham/uniswap-hardcoded`, `QmdmdZRG42GG4C4JWNCnf5uwGAG8Uh6Q6iAM1VDLnWKNzq`) which weighs in at around 50GB, and the deploy took about 4 hours and 45 minutes on a test database. The test database is much slower than prod, but even in prod this would be a significant delay.

The bulk of the time for copying is taken up by building indexes: I just reran grafting the above subgraph, once with attribute indexes created before data was copied, and once with copying the data first and creating the attribute indexes afterwards. In the first case, grafting took 280 minutes (4h 40mins), in the second it took 270 minutes (4h 30mins). The actual data copying in the second case took roughly 30 minutes, and the remaining 4 hours were taken up by index building.

Of course, this is a somewhat extreme case, and smaller subgraphs are much faster to copy.

**Comment by lutter** [comment]
Rebased onto latest master; add reference docs for `graft` in subgraph manifest

File: store/postgres/src/relational_queries.rs [hunk 4]
```diff
@@ -2173,3 +2173,117 @@ impl<'a> LoadQuery<PgConnection, RevertEntityData> for DeleteByPrefixQuery<'a> {
...
+
+impl<'a> QueryFragment<Pg> for CopyDynamicDataSourceQuery<'a> {
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        // The query in this file was generated with
+        //    cargo run -p graph-store-postgres --example layout -- \
+        //          -g copy-dds store/postgres/src/subgraphs.graphql \
+        //          > store/postgres/src/copy_dds.sql
+        // Background on how the query works can be found in layout.rs
+        // in the function `print_copy_dds`
+        // See also: ed42d219c6704a4aab57ce1ea66698e7
```
Comment (Jannis) [inline store/postgres/src/relational_queries.rs:2271]: Ah, great, here is the note I was asking about.

File: store/postgres/src/store.rs [hunk 5]
```diff
@@ -1016,6 +967,21 @@ impl StoreTrait for Store {
         if block_ptr_from.number != block_ptr_to.number + 1 {
             panic!("revert_block_operations must revert a single block only");
         }
+        // Don't revert past a graft point
+        let info = self.subgraph_info(&subgraph_id)?;
+        if let Some(graft_block) = info.graft_block {
+            if graft_block as u64 > block_ptr_to.number {
+                return Err(format_err!(
+                    "Can not revert subgraph `{}` to block {} \
+                  as it was grafted at block {} and we do not allow reverting \
```
Comment (Jannis) [inline store/postgres/src/store.rs:976]: Let's rephrase the `we do not allow` to `reverting past graft points is not allowed` or something else that does not involve "us".

File: core/src/subgraph/registrar.rs (hunk 1 above)
Comment (lutter) [reply core/src/subgraph/registrar.rs:482]: I wasn't planning on doing that in this PR ;)

File: graph/src/data/subgraph/mod.rs [hunk 6]
```diff
@@ -844,6 +846,42 @@ impl DataSourceTemplate {
...
+    ) -> Vec<SubgraphManifestValidationError> {
+        fn gbi(msg: String) -> Vec<SubgraphManifestValidationError> {
+            vec![SubgraphManifestValidationError::GraftBaseInvalid(msg)]
+        }
+
+        match store.block_ptr(self.base.clone()) {
+            Err(e) => gbi(e.to_string()),
+            Ok(None) => gbi(format!(
+                "can not graft onto `{}` since it has not processed any blocks",
+                self.block
```
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/data/subgraph/mod.rs:869]: Yep. Fixed that

File: graph/src/data/subgraph/mod.rs [hunk 7]
```diff
@@ -844,6 +846,42 @@ impl DataSourceTemplate {
...
+        store: Arc<S>,
+    ) -> Vec<SubgraphManifestValidationError> {
+        fn gbi(msg: String) -> Vec<SubgraphManifestValidationError> {
+            vec![SubgraphManifestValidationError::GraftBaseInvalid(msg)]
+        }
+
+        match store.block_ptr(self.base.clone()) {
+            Err(e) => gbi(e.to_string()),
+            Ok(None) => gbi(format!(
+                "can not graft onto `{}` since it has not processed any blocks",
```
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/data/subgraph/mod.rs:868]: Changed to `failed to`

File: store/postgres/src/copy_dds.sql [hunk 8]
```diff
@@ -0,0 +1,60 @@
+with xlat as (
+  select * from unnest($1::text[], $2::text[]) as xlat(id, new_id)),
+ md0 as (
+    insert into subgraphs.ethereum_block_handler_entity(id, handler, filter, block_range)
+    select (x.new_id || right(e.id, -40)) as id, handler, (x.new_id || right(e.filter, -40)) as filter, block_range
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/copy_dds.sql:5]: `||` in SQL is string concatentation, so we are concatenating `x.new_id` with whatever is in `e.id` past the first 40 chars (`right(e.id, -40)` gives you everything but the first 40 chars) In other words, we are replacing the first 40 characters of `e.id` with `x.new_id`

File: store/postgres/src/copy_dds.sql [hunk 9]
```diff
@@ -0,0 +1,60 @@
+with xlat as (
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/copy_dds.sql:1]: Yes it will have to be regenerated when the metadata schema changes. I could easily turn this into something that is dynamically generated to make that unnecessary. I wasn't sure whether it's preferrable to have to regenerate, but be able to inspect the query, or to not have to regenerate and leave the query harder to get to.

File: store/postgres/src/entities.rs [hunk 10]
```diff
@@ -340,8 +342,60 @@ impl Connection {
...
+                            unreachable!("A JSONB subgraph is never grafted onto another subgraph");
+                        }
+                        diesel::update(dsl::table)
+                            .set(dsl::state.eq(State::Ready))
+                            .filter(dsl::subgraph.eq(self.storage.subgraph().as_str()))
+                            .execute(&self.conn)?;
+                    }
+                }
+            }
+            State::Tables => unimplemented!("continue the migration"),
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:395]: We would only hit this if we ran a migration through the 'migrate one subgraph at a time' machinery that we used to go from JSONB to relational storage, and then only if a server is shut down in the middle of such a migration. We don't have any migrations using these mechanisms, so can't hit that right now.

File: store/postgres/src/entities.rs [hunk 11]
```diff
@@ -792,14 +841,50 @@ impl Connection {
...
+                    &schema.document,
+                )?;
+                // See if we are grafting and check that the graft is permissible
+                if let Some((base, _)) = metadata::deployment_graft(&self.conn, &schema.id)? {
+                    match Storage::new(&self.conn, &base)? {
+                        Storage::Relational(base) => {
+                            let errors = layout.can_copy_from(&base);
+                            if !errors.is_empty() {
+                                return Err(StoreError::Unknown(format_err!(
+                                    "The subgraph `{}` can not be used as the graft base \
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:858]: Changed.

File: store/postgres/src/entities.rs [hunk 12]
```diff
@@ -792,14 +841,50 @@ impl Connection {
...
+                                        for `{}` because the schemas are incompatible:\n    - {}",
+                                    &base.subgraph,
+                                    &layout.subgraph,
+                                    errors.join("\n    - ")
+                                )));
+                            }
+                        }
+                        Storage::Json(json) => {
+                            return Err(StoreError::Unknown(format_err!(
+                                "The subgraph `{}` can not be used as the graft base \
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:868]: Changed

File: store/postgres/src/relational.rs [hunk 13]
```diff
@@ -251,6 +252,87 @@ impl Layout {
...
+        block: EthereumBlockPointer,
+        metadata: &Layout,
+    ) -> Result<(), StoreError> {
+        // This can not be used to copy data to or from the metadata subgraph
+        assert!(!self.subgraph.is_meta());
+        assert!(!base.subgraph.is_meta());
+
+        info!(
+            logger,
+            "Initializing graft by copying data from {} to {}", base.subgraph, self.subgraph
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/relational.rs:269]: The `graft_target` is already logged as the `subgraph_id` of the logger that is used here and I have been using that to extract logs of subgraph copying. What you get is something like
```
Apr 10 16:16:44.450 INFO Initializing graft by copying data from QmYzFwdV58HiZX5fJo6WoRXjnnzvWzzvpcCcLreeifTPZR to QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, subgraph_id: QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, component: SubgraphInstanceManager
Apr 10 16:16:59.615 INFO Copied 1272984 UserExchangeData entities, time_ms: 15165, subgraph_id: QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, component: SubgraphInstanceManager
Apr 10 16:17:07.575 INFO Copied 1128193 ExchangeDayData entities, time_ms: 7959, subgraph_id: QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, component: SubgraphInstanceManager
...
Apr 10 16:57:40.294 INFO Copied 2760 dynamic data sources, time_ms: 60931, subgraph_id: QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, component: SubgraphInstanceManager
Apr 10 17:04:02.619 INFO Rewound subgraph to block 9467000, time_ms: 382324, subgraph_id: QmUxERojju4NxDN3FgY6JFeh1xumT32TL81jZUViSXeWay, component: SubgraphInstanceManager
```

I feel that that is good enough for focusing on logging - I doubt searching by `graft_base` will ever be useful.

File: store/postgres/src/relational.rs [hunk 14]
```diff
@@ -251,6 +252,87 @@ impl Layout {
...
+        let new_dds = (0..dds.len())
+            .map(|_| DynamicEthereumContractDataSourceEntity::make_id())
+            .collect::<Vec<_>>();
+        // Copy the data sources and all their subordinate entities, translating
+        // ids into new ids in the process and attaching them to `self.subgraph`
+        rq::CopyDynamicDataSourceQuery::new(&dds, &new_dds, self.subgraph.as_str())
+            .execute(conn)?;
+        info!(logger, "Copied {} dynamic data sources", dds.len();
+              "time_ms" => start.elapsed().as_millis());
+        // 3. Rewind the subgraph. `revert_block` gets rid of everything
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/relational.rs:305]: Changed

File: store/postgres/src/relational.rs [hunk 15]
```diff
@@ -251,6 +252,87 @@ impl Layout {
...
+    ) -> Result<(), StoreError> {
+        // This can not be used to copy data to or from the metadata subgraph
+        assert!(!self.subgraph.is_meta());
+        assert!(!base.subgraph.is_meta());
+
+        info!(
+            logger,
+            "Initializing graft by copying data from {} to {}", base.subgraph, self.subgraph
+        );
+        // 1. Copy subgraph data
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/relational.rs:271]: Changed

File: store/postgres/src/relational.rs [hunk 16]
```diff
@@ -1034,6 +1203,69 @@ mod tests {
...
+        // We can not change the underlying type of a field in arrays
+        let source = test_layout("type Scalar { id: ID, color: [Int!]! }");
+        let dest = test_layout("type Scalar { id: ID, color: [String!]! }");
+        assert_eq!(
+            vec![
+                "The attribute Scalar.color has type [String!]!, but \
+                 its type in the source is [Int!]!"
+            ],
+            dest.can_copy_from(&source)
+        );
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/relational.rs:1266]: Truth be told, and AFAICT, we don't really distinguish between arrays with and without nulls when storing entities (and I don't think we ever have).

File: store/postgres/src/relational.rs [hunk 17]
```diff
@@ -1034,6 +1203,69 @@ mod tests {
...
+    }
+
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/relational.rs:1268]: Yes, that's what `EnumType.is_assignable_from` is about: the target enum type must contain all the values of the source type, but it's totally fine to have more.

File: store/postgres/src/store.rs (hunk 5 above)
Comment (lutter) [reply store/postgres/src/store.rs:976]: I changed it to `and reverting past a graft point is not possible`

**Comment by lutter** [comment]
Addressed all review comments

File: store/postgres/src/copy_dds.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/copy_dds.sql:5]: Ah, cool, so just swapping out the old subgraph ID (`Qm...`) but not the suffix that makes the handler entity ID unique.

File: store/postgres/src/entities.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/entities.rs:395]: Ok

File: store/postgres/src/relational.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/relational.rs:269]: Ok, fair enough, that does look pretty clear.

File: store/postgres/src/relational.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/relational.rs:1268]: True

**Comment by lutter** [comment]
Rebased to latest master

File: store/postgres/src/relational.rs [hunk 18]
```diff
@@ -251,6 +252,89 @@ impl Layout {
...
+        // 3. Rewind the subgraph. `revert_block` gets rid of everything
+        // including the block passed to it. We want to preserve `block`
+        // and therefore revert `block+1`
+        let start = Instant::now();
+        let revert_to: BlockNumber = (block.number + 1)
+            .try_into()
+            .expect("block numbers fit into an i32");
+        self.revert_block(conn, revert_to)?;
+        metadata.revert_metadata(conn, &self.subgraph, revert_to)?;
```
Comment (Jannis) [inline store/postgres/src/relational.rs:315]: Is `revert_to` the right name if we actually revert to `revert_to - 1`?

File: store/postgres/src/relational.rs [hunk 19]
```diff
@@ -631,6 +713,28 @@ impl Layout {
...
+        if source.values.is_subset(self.values.as_ref()) {
+            None
+        } else {
+            Some(format!(
+                "the enum type {} contains values not present in {}",
+                source.name, self.name
+            ))
+        }
+    }
+}
```
Comment (Jannis) [inline store/postgres/src/relational.rs:736]: The more I look at this, the more I think these checks should take place at the GraphQL schema level. That's intended to be the source of truth, and the Postgres schema is just how it's manifested at the storage layer.

File: store/postgres/src/entities.rs [hunk 20]
```diff
@@ -340,8 +342,60 @@ impl Connection {
...
+            .first::<State>(&self.conn)?;
+
+        match state {
+            State::Init => {
+                let graft = metadata::deployment_graft(&self.conn, &self.storage.subgraph())?;
+                match &*self.storage {
+                    Storage::Relational(layout) => {
+                        let start = Instant::now();
+                        let graft =
+                            metadata::deployment_graft(&self.conn, &self.storage.subgraph())?;
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:361]: Copypasta. Fixed

File: store/postgres/src/entities.rs [hunk 21]
```diff
@@ -792,14 +841,50 @@ impl Connection {
...
+                    &self.conn,
+                    &schema_name,
+                    schema.id.clone(),
+                    &schema.document,
+                )?;
+                // See if we are grafting and check that the graft is permissible
+                if let Some((base, _)) = metadata::deployment_graft(&self.conn, &schema.id)? {
+                    match Storage::new(&self.conn, &base)? {
+                        Storage::Relational(base) => {
+                            let errors = layout.can_copy_from(&base);
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:855]: I'd be loathe to do it with the tools we have for handling GraphQL schemas right now. A big difference between a `Layout` and the GraphQL `Document` is that a lot of potential problems are clearly not possible anymore with a `Layout`: for example, when you have a `Layout`, you know that each attribute has a valid type. When you deal with a GraphQL `Document` every single time you need to look at the type of an attribute, you have to be prepared for the fact that the user might have typoed `Account` as `Accuont` Even if the code flow says that it was checked before, is it really ok to use `unwrap` or `expect` in all those cases? It embeds a lot of knowledge along the lines of 'the data structure makes this possible, but we know from how the code flows that it's really not a problem here' and results in GraphQL schema validation code that is very hard to follow.

File: store/postgres/src/entities.rs [hunk 22]
```diff
@@ -792,14 +841,50 @@ impl Connection {
...
-            )
-            .map(|_| ()),
-            v::Split => create_split_schema(&self.conn, &schema_name),
+            v::Relational => {
+                let layout = Layout::create_relational_schema(
+                    &self.conn,
+                    &schema_name,
+                    schema.id.clone(),
+                    &schema.document,
+                )?;
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/entities.rs:850]: All this is happening inside one transaction. Postgres (unlike a lot of other databases) makes DDL transactional, meaning if the transaction gets rolled back because of an error farther down the line, all the DDL statements we issued get rolled back, too.

File: store/postgres/src/metadata.rs [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+        use subgraph_manifest as sm;
+        let manifest_id = SubgraphManifestEntity::id(&id);
+        let s: String = sm::table
+            .select(sm::schema)
+            .filter(sm::id.eq(manifest_id.as_str()))
+            .first(conn)?;
+        Schema::parse(s.as_str(), id)
+    };
+    res.map_err(|e| StoreError::Unknown(e))
+}
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/metadata.rs:139]: It's so you can get at these things without needing a `Store`. The way it was before introduced a circular dependency where a `Store` needed an `entities::Connection`, and to construct an `entities::Connection` you needed a `Store`. The latter was solely because the `Store` is also a cache of schemas. 

That leads to two complications: (1) we now want the `Store` to also cache information about a possible graft point since that needs to be checked on every revert; when we create a grafted subgraph there's a point where we need an `entities::Connection` before the graft information has been written to the database. If we go through the cache, we could end up storing incorrect information in the cache, or need to remember to invalidate the cache entry in certain cases. It's more robust to have a way to get the schema that bypasses the cache when we are not sure things are truly cacheable yet. (2) It causes trouble with @That3Percent's `with_entity_connection` API for similar reasons (i.e., the seeming circularity of the `Store` and `entities::Connection` dependencies because of caching)

File: store/postgres/src/relational.rs (hunk 18 above)
Comment (lutter) [reply store/postgres/src/relational.rs:315]: Hard to come up with a good name here (hence the comment just before this) I changed it to `block_to_revert`

File: store/postgres/src/relational.rs (hunk 19 above)
Comment (lutter) [reply store/postgres/src/relational.rs:736]: The problem I have with that is that it is not apparent at all what the truth is when you have a `Schema` object (with the parser's raw `Document` in it) Even if we decide it needs to go to the GraphQL layer right now, I'd prefer to do that as a follow-on PR rather than leaving this open until that's done.

**Comment by lutter** [comment]
Addressed the directly actionable review comments

File: store/postgres/src/entities.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/entities.rs:853]: Yeah, I know that the raw GraphQL schema is awkward to work with. At the same time, decisions like whether two schemas are compatible should not be made deep down in the store either. The fact that even Graph CLI could check for schema compatibility also speaks against tying it so closely to the store implementation.

It could be nice if we translated the unvalidated schema `Document` into a validated `Schema` type that has an unambiguous, checked and convenient shape. The translation from that to a `Layout` should be straight-forward, with no potential for errors.

But... that's for another day.

File: store/postgres/src/entities.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/entities.rs:848]: Sorry, I now see this is running inside `econn.transaction(...)`, which of course means everything's rolled back in case of an error. Nevermind 😄

File: store/postgres/src/metadata.rs [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+pub fn subgraph_schema(
+    conn: &PgConnection,
+    id: SubgraphDeploymentId,
+) -> Result<Schema, StoreError> {
+    // The subgraph of subgraphs schema is built-in and doesn't have a
+    // SubgraphManifest in the database
+    const SUBGRAPHS_SCHEMA: &str = include_str!("subgraphs.graphql");
+    let res = if id.is_meta() {
+        Schema::parse(SUBGRAPHS_SCHEMA, id)
```
Comment (Jannis) [inline store/postgres/src/metadata.rs:128]: I wonder what's cheaper: reparsing this schema every time we access the metadata subgraph, or making it `lazy_static!`, parsing it once, and cloning it every time...

File: store/postgres/src/metadata.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply store/postgres/src/metadata.rs:139]: Ok, got it!

File: store/postgres/src/relational.rs (hunk 18 above)
Comment (Jannis) [reply store/postgres/src/relational.rs]: That works 😄

File: store/postgres/src/relational.rs (hunk 19 above)
Comment (Jannis) [reply store/postgres/src/relational.rs:736]: Yeah, agreed (see the other comment).

File: store/postgres/src/metadata.rs (hunk 24 above)
Comment (lutter) [reply store/postgres/src/metadata.rs:128]: Yeah, good point. Ideally, this could return an `Arc<Schema>` It's a little tricky right now, but I think once my other PR gets in, it would be easy to do that and then just keep the metadata subgraph in a `lazy_static!` `Arc`

**Comment by lutter** [comment]
Rebased to latest master
